modules/block-interactions/src/main/java/unnamed/mmo/blocks/ore/handler/OreBlockHandler.java

```diff
@@ -0,0 +1,140 @@
+package unnamed.mmo.blocks.ore.handler;
+
+import net.minestom.server.coordinate.Point;
+import net.minestom.server.entity.Player;
+import net.minestom.server.instance.Instance;
+import net.minestom.server.instance.block.Block;
+import net.minestom.server.instance.block.BlockHandler;
+import net.minestom.server.utils.NamespaceID;
+import net.minestom.server.utils.validate.Check;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import unnamed.mmo.blocks.ore.Ore;
+import unnamed.mmo.loot.LootContext;
+import unnamed.mmo.server.instance.TickTrackingInstance;
+import unnamed.mmo.util.BlockUtil;
+
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * Stateless singleton block handler for all ore blocks.
+ */
+public class OreBlockHandler implements BlockHandler {
+    private static final OreBlockHandler instance = new OreBlockHandler();
+
+    public static @NotNull OreBlockHandler instance() {
+        return instance;
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(OreBlockHandler.class);
+
+    //todo this should probably be configurable.
+    private static final Block REPLACEMENT_BLOCK = Block.BEDROCK;
+
+    private OreBlockHandler() {
+        Check.stateCondition(instance != null, "There can only be one OreBlockHandler");
+    }
+
+    @Override
+    public void onPlace(@NotNull Placement placement) {
+        final Instance instance = placement.getInstance();
+        final Point pos = placement.getBlockPosition();
+        final Block block = placement.getBlock();
+
+        // Ensure the block is an ore
+        final Ore ore = Ore.fromBlock(block);
+        if (ore == null) {
+            // Block handler should not be present if the ore tag is not present. Remove it and log error.
+            LOGGER.error("Handler present on a block without ore tag: {} at {} in {}", block, pos, instance);
+            instance.setBlock(pos, block.withHandler(null));
+            return;
+        }
+
+        // Do nothing if the block is the replacement
+        if (block.compare(REPLACEMENT_BLOCK, Block.Comparator.STATE))
+            return;
+
+        // Ensure the block has the right state
+        if (block.compare(ore.oreBlock(), Block.Comparator.STATE))
+            return;
+
+        // Block should already be configured, if not we have a bug.
+        LOGGER.error("Ore block has wrong state: {} at {} in {}", block, pos, instance);
+        instance.setBlock(pos, BlockUtil.withType(block, ore.oreBlock()));
+    }
+
+    @Override
+    public void onDestroy(@NotNull Destroy destroy) {
+        final Block block = destroy.getBlock();
+        if (block.compare(REPLACEMENT_BLOCK, Block.Comparator.STATE) || !(destroy instanceof PlayerDestroy))
+            return;
+
+        final Instance instance = destroy.getInstance();
+        final Point pos = destroy.getBlockPosition();
+        final Ore ore = Ore.fromBlock(block);
```

Comment:
Duplicated code from onPlace, perhaps put into a separate checker method?

Reply:
I tried this, but the method needs so many parameters and the if statement still has to be there so it doesn't seem like its an improvement to me.